From 73cf70423a3b0887e654925507d7928f83513fb3 Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Fri, 9 Oct 2015 18:54:57 -0700
Subject: [PATCH 0850/1767] UBUNTU: SAUCE: apparmor: Fix: ensure aa_get_newest
 will trip debugging if the replacedby is not setup

BugLink: http://bugs.launchpad.net/bugs/1448912

While we are at it, also catch any potential refcounting errors arround
the replacedby label dereference.

Signed-off-by: John Johansen <john.johansen@canonical.com>
Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---
 security/apparmor/include/label.h | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/security/apparmor/include/label.h b/security/apparmor/include/label.h
index 071eabe..3530efe 100644
--- a/security/apparmor/include/label.h
+++ b/security/apparmor/include/label.h
@@ -384,8 +384,17 @@ static inline struct aa_label *aa_get_newest_label(struct aa_label *l)
 	if (!l)
 		return NULL;
 
-	if (label_invalid(l))
-		return aa_get_label_rcu(&l->replacedby->label);
+	if (label_invalid(l)) {
+		struct aa_label *tmp;
+		AA_BUG(!l->replacedby);
+		AA_BUG(!l->replacedby->label);
+		/* BUG: only way this can happen is @l ref count and its
+		 * replacement count have gone to 0 and are on their way
+		 * to destruction. ie. we have a refcounting error
+		 */
+		AA_BUG(!(tmp = aa_get_label_rcu(&l->replacedby->label)));
+		return tmp;
+	}
 
 	return aa_get_label(l);
 }
-- 
1.9.3

